datastore/longhorn.go

```diff
@@ -3429,3 +3455,167 @@ func (s *DataStore) ListOrphansByNodeRO(name string) ([]*longhorn.Orphan, error)
 func (s *DataStore) DeleteOrphan(orphanName string) error {
 	return s.lhClient.LonghornV1beta2().Orphans(s.namespace).Delete(context.TODO(), orphanName, metav1.DeleteOptions{})
 }
+
+type Disk struct {
+	spec   *longhorn.DiskSpec
+	status *longhorn.DiskStatus
+}
+
+func (s *DataStore) IsStorageSpaceSufficientForSnapshotPurge(volumeName string) (bool, error) {
+	replicas, err := s.ListReplicasByVolume(volumeName)
+	if err != nil {
+		return false, err
+	}
+
+	disks, err := s.listDisksContainingReplicas(replicas)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to list disks containing replicas for volume %v", volumeName)
+	}
+
+	requiredPurgeReservedSpace, err := s.getRequiredPurgeReservedSpace(volumeName)
```

Comment:
Yes, I count the purging replicas in `countExistingPurgeReservedSizeOnDisk()`.
`getRequiredPurgeReservedSpace()` only get the reserved space of this volume.
Overall reserved space is `countExistingPurgeReservedSizeOnDisk()+ getRequiredPurgeReservedSpace ()`.

Reply:
Oh my bad, didn't notice that you already add it.
In the future, we can enhance this part a little bit to avoid overestimating the reversed space:
1. For the snapshot being marked as removed, we know that the extra space is the actual size of its child * (100% - PurgeProgress) when trying to purge it.
2. For the snapshot directly behind the volume head, there is no need to consider the extra space.

Reply:
Good idea. I will enhance this part.

Reply:
Just a note. This is quite important to have 2 items included, especially since we will enforcedly avoid users do the snapshot purge.  

Thanks @shuo-wu and @derekbit

Reply:
@shuo-wu 
One question, how do we know if the snapshot to be purged is behind the volume head?

Reply:
Check the parent of the snapshot is volume-head and it's also purging?

Reply:
Yes, just think of the engine CR has snapshot chain, so we can walk through the snapshots marked deleted.

Reply:
Each snapshot contains the children's info. Then we can utilize it to avoid reserved space overestimation. You already notice that we cannot directly get the following 2 important info there:
    1. which removed snapshot is being purged 
    2. the actual op of the purge. 
    3. How much extra space Longhorn still needs to reserve for the current puring.
For the 1st, I have no idea for detecting it. But for the 2nd one, I can deduce it from the snapshot chain/snapshot children info. As for the 3rd one, we can roughly say that the extra space required = (the largest child snapshot among all removing snapshot) * (100% - CurrentPurgeProgress)